src/charm.py

```diff
@@ -9,19 +9,133 @@
 """
 
 import logging
+import os
 from typing import Any
 
 import ops
+import yaml
+from ops import CollectStatusEvent, ConfigChangedEvent, RelationChangedEvent
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
 
 logger = logging.getLogger(__name__)
 
+CLOUDS_YAML = "/etc/openstack/clouds.yaml"
+# used to match entry in clouds.yaml with openstack-exporter config
+CLOUD_NAME = "openstack"
+
 
 class OpenstackExporterOperatorCharm(ops.CharmBase):
     """Charm the service."""
 
     def __init__(self, *args: tuple[Any]) -> None:
         """Initialize the charm."""
         super().__init__(*args)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.credentials_relation_changed, self._on_credentials_changed)
+        self.framework.observe(self.on.collect_unit_status, self._on_collect_unit_status)
+
+    def _on_credentials_changed(self, _: RelationChangedEvent) -> None:
+        """Handle updates to credentials from keystone."""
+        self._configure()
+
+    def _on_config_changed(self, _: ConfigChangedEvent) -> None:
+        """Reconfigure everything on config changed."""
+        self._configure()
+
+    def _configure(self) -> None:
+        """Reconfigure everything.
+
+        This is an idempotent method to configure everything.
+        """
+        data = self._get_keystone_data()
+
+        if data:
+            logger.info("Keystone credentials are available, starting services.")
+            self.write_cloud_config(data)
+            # TODO: when openstack-exporter service is added, restart it here
+        else:
+            logger.info("Keystone credentials are not available, stopping services.")
```

Comment:
Do we need to defer here? I can imagine if we can't get the relation data the first try when credentials changed, then we need to remove-relation and relate again

Reply:
Okay, thanks